nav2_waypoint_follower/plugins/photo_at_waypoint.cpp

```diff
@@ -0,0 +1,158 @@
+// Copyright (c) 2020 Fetullah Atas
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "nav2_waypoint_follower/plugins/photo_at_waypoint.hpp"
+#include <pluginlib/class_list_macros.hpp>
+
+#include <string>
+#include <memory>
+
+
+namespace nav2_waypoint_follower
+{
+PhotoAtWaypoint::PhotoAtWaypoint()
+{
+}
+
+PhotoAtWaypoint::~PhotoAtWaypoint()
+{
+}
+
+void PhotoAtWaypoint::initialize(
+  const rclcpp_lifecycle::LifecycleNode::WeakPtr & parent,
+  const std::string & plugin_name)
+{
+  auto node = parent.lock();
+
+  node->declare_parameter(plugin_name + ".enabled", rclcpp::ParameterValue(true));
+  node->declare_parameter(
+    plugin_name + ".image_topic",
+    rclcpp::ParameterValue("/camera/color/image_raw"));
+  node->declare_parameter(
+    plugin_name + ".save_dir",
+    rclcpp::ParameterValue("/tmp/waypoint_images"));
+  node->declare_parameter(plugin_name + ".image_format", rclcpp::ParameterValue("png"));
+
+  node->get_parameter(plugin_name + ".enabled", is_enabled_);
+  node->get_parameter(plugin_name + ".image_topic", image_topic_);
+  node->get_parameter(plugin_name + ".save_dir", directory_to_save_images_);
+  node->get_parameter(plugin_name + ".image_format", image_format_);
+
+  if (!is_enabled_) {
+    RCLCPP_INFO(
+      logger_, "Waypoint task executor plugin is disabled.");
+  } else {
+    RCLCPP_INFO(
+      logger_, "Initializing photo at waypoint plugin, subscribing to camera topic named; %s",
+      image_topic_.c_str());
+    camera_image_subscriber_ = node->create_subscription<sensor_msgs::msg::Image>(
+      image_topic_, rclcpp::SystemDefaultsQoS(),
+      std::bind(&PhotoAtWaypoint::imageCallback, this, std::placeholders::_1));
+  }
+}
+
+bool PhotoAtWaypoint::processAtWaypoint(
```

Comment:
Have you run tests to make sure this all works as expected and images are saved where you think they should be?

Reply:
yes, here is a video where i do shelf inspection with this plugin;
https://www.youtube.com/watch?v=-fJ2g46BPdo 
The video is quite long (4 mins) , jump to minute 3 where it will display the taken photos by the plugin

Reply:
Even if you can't do it in the next day, it would still be useful to have done for the tutorial either way!

Reply:
No problem, at moment I cant do with the real robot though, only simulation

Reply:
Here is 4K video, it is 4X speed otherwise it would take more than a minute, 
https://www.youtube.com/watch?v=NLzk1TcQUXQ 
Please mention about the affiliation and author as; 
Chicony Electronics, Fetullah Atas